File: src/classes/join/join_req.rs
use crate::{consts::{other::Other, errors::HbrsError}, utils::byte_arrays_helper::ByteArraysHelper};

#[derive(Debug, Clone)]
pub struct JoinReq{
    ssid: Vec<u8>,
    security_type: u8,
    encryption_type: u8,
    key: Vec<u8>,
}

impl JoinReq{
    pub fn new(ssid: Vec<u8>, security_type: u8, encryption_type: u8, key: Vec<u8>) -> Result<Self, HbrsError>{
        let key = match ByteArraysHelper::cp_arr_bytes_fill(&key, Other::KeyLength.get_value()){
            Ok(value) => value,
            Err(err) => return Err(err),
        };

        Ok(Self{
            ssid: ssid,
            security_type: security_type,
            encryption_type: encryption_type,
            key: key,
        })
    }

    pub fn get_bytes(&self) -> Vec<u8>{
        let first = ByteArraysHelper::combine_1v_1b(&self.ssid, self.security_type);
        let second = ByteArraysHelper::combine_1v_1b(&self.key, self.encryption_type);
        return ByteArraysHelper::combine_2v(&first, &second);
    }

    pub fn get_ssid_ref(&self) -> &Vec<u8> {
        &self.ssid
    }
    
    pub fn get_ssid_mut(&mut self) -> &mut Vec<u8> {
        &mut self.ssid
    }
    
    pub fn get_security_type_ref(&self) -> &u8 {
        &self.security_type
    }
    
    pub fn get_security_type_mut(&mut self) -> &mut u8 {
        &mut self.security_type
    }
    
    pub fn get_encryption_type_ref(&self) -> &u8 {
        &self.encryption_type
    }
    
    pub fn get_encryption_type_mut(&mut self) -> &mut u8 {
        &mut self.encryption_type
    }
    
    pub fn get_key_ref(&self) -> &Vec<u8> {
        &self.key
    }
    
    pub fn set_key(&mut self, key: Vec<u8>) -> Result<(), HbrsError>{
        self.key = match ByteArraysHelper::cp_arr_bytes_fill(&key, Other::KeyLength.get_value()){
            Ok(value) => value,
            Err(err) => return Err(err),
        };

        Ok(())
    }
}

File: lib/classes/join/join_req.py
from lib.consts.other import Other
from lib.utils.byte_arrays_helper import ByteArraysHelper


class JoinReq:
    def __init__(self, ssid: list[int], security_type: int, encryption_type: int, key: list[int]):
        self.ssid = ssid
        self.security_type = security_type
        self.encryption_type = encryption_type
        self.key = key

    @staticmethod
    def new(ssid: list[int], security_type: int, encryption_type: int, key: list[int]):
        key = ByteArraysHelper.cp_arr_bytes_fill(key, Other.KeyLength)
        return JoinReq(ssid, security_type, encryption_type, key)

    def get_bytes(self) -> list[int]:
        first = ByteArraysHelper.combine_1v_1b(self.ssid, self.security_type)
        second = ByteArraysHelper.combine_1v_1b(self.key, self.encryption_type)
        return ByteArraysHelper.combine_2v(first, second)
    
    def set_key(self, key: list[int]) -> None:
        self.key = ByteArraysHelper.cp_arr_bytes_fill(key, Other.KeyLength)

    def get_ssid(self):
        return self.ssid

    def set_ssid(self, ssid: list[int]):
        self.ssid = ssid

    def get_security_type(self):
        return self.security_type

    def set_security_type(self, security_type: int):
        self.security_type = security_type

    def get_encryption_type(self):
        return self.encryption_type

    def set_encryption_type(self, encryption_type: int):
        self.encryption_type = encryption_type

    def get_key(self):
        return self.key

    def set_key(self, key: list[int]):
        self.key = key
